temporal-serviceclient/src/main/java/io/temporal/internal/BackoffThrottler.java

```diff
@@ -60,32 +71,58 @@
 @NotThreadSafe
 public final class BackoffThrottler {
 
-  private final Duration initialSleep;
+  private final Duration regularInitialSleep;
+
+  private final Duration congestionInitialSleep;
 
   private final Duration maxSleep;
 
   private final double backoffCoefficient;
 
+  private final double maxJitter;
+
   private int failureCount = 0;
 
+  private Status.Code lastFailureCode = Code.OK;
+
   /**
    * Construct an instance of the throttler.
    *
-   * @param initialSleep time to sleep on the first failure
+   * @param regularInitialSleep time to sleep on the first failure (assuming regular failures)
+   * @param congestionInitialSleep time to sleep on the first failure (for congestion failures)
    * @param maxSleep maximum time to sleep independently of number of failures
-   * @param backoffCoefficient coefficient used to calculate the next time to sleep.
+   * @param backoffCoefficient coefficient used to calculate the next time to sleep
+   * @param maxJitter maximum jitter to add or subtract
    */
   public BackoffThrottler(
-      Duration initialSleep, @Nullable Duration maxSleep, double backoffCoefficient) {
-    Objects.requireNonNull(initialSleep, "initialSleep");
-    this.initialSleep = initialSleep;
+      Duration regularInitialSleep,
+      Duration congestionInitialSleep,
+      @Nullable Duration maxSleep,
+      double backoffCoefficient,
+      double maxJitter) {
+    Objects.requireNonNull(regularInitialSleep, "regularInitialSleep");
+    Objects.requireNonNull(congestionInitialSleep, "congestionInitialSleep");
+    if (backoffCoefficient < 1.0) {
+      throw new IllegalArgumentException(
+          "backoff coefficient less than 1.0: " + backoffCoefficient);
+    }
+    if (maxJitter < 0 || maxJitter >= 1.0) {
+      throw new IllegalArgumentException("maximumJitter has to be >= 0 and < 1.0: " + maxJitter);
+    }
+    this.regularInitialSleep = regularInitialSleep;
+    this.congestionInitialSleep = congestionInitialSleep;
     this.maxSleep = maxSleep;
     this.backoffCoefficient = backoffCoefficient;
+    this.maxJitter = maxJitter;
   }
 
   public long getSleepTime() {
     if (failureCount == 0) return 0;
-    double sleepMillis = Math.pow(backoffCoefficient, failureCount - 1) * initialSleep.toMillis();
+    Duration initial =
+        (lastFailureCode == Code.RESOURCE_EXHAUSTED) ? congestionInitialSleep : regularInitialSleep;
+    double jitter = Math.random() * maxJitter * 2 - maxJitter;
```

Comment:
I would add a comment here that "we generate a randomize jitter coefficient in (-1, 1) range"

Reply:
Jitter is not the most pleasant mechanic for a user to understand what's going on. I would keep it effective only for codes where jitter actually physically makes sense.
Jitter is needed for Code.RESOURCE_EXHAUSTED to avoid situations when all the parties fail and retry at the same time, getting an uneven throttle. For other codes... probably not?

Reply:
>Jitter is not the most pleasant mechanic for a user to understand what's going on.

Can you elaborate on what is your concern here, from user's POV? Exact timing in unit tests?

In most cases, for non-congestion errors, I'd expect that the numbers we are talking here should be less that some other external factors that might add delay (DNS lookup, network level congestion, garbage collection...). Assuming defaults values, jitter would add at most +/- 20 ms on the third attempt (i.e. effective sleep time in the range [180, 220] ms). Does that sounds bad?

Reply:
> Can you elaborate on what is your concern here, from user's POV? Exact timing in unit tests?

Users may be surprised by unexpected unstable retry periods and may percieve it as a problem/bug. Stable retry periods are easier to grasp and more standard.

> Does that sounds bad?
It's not bad, it's just more complicated than it has to be.

How is it done in other SDKs? Do we apply jitter for all failures?

Reply:
TypeScript applies a jitter of 10% by default (user configurable) on _every_ grpc failures.

Go applies a jitter of 20% (non configurable) on _every_ grpc failures. The formula is little bit different though (actual sleep time is random in the range `[ sleep before jitter * 0.8, sleep before jitter ]`)